### PR TITLE
Refac field dimensions

### LINKED ISSUE #415: Split fieldDimensions into horizontal & vertical
We need to prepare the toolchain to support unstructured masked dimensions. To do that, we should split the "fieldDimensions" (a tuple of booleans) into horizontal and vertical "field dimensions", similarly as we did it for extents and offsets.
- Create a `CartesianFieldDimension` class and a `FieldDimensions` class
- No `UnstructuredFieldDimension` yet
- Define, document and test the proper interface to support all use-cases in the code.

### PR DESCRIPTION
**Pull request by mroethlin**
## Technical Description

Splits the "fieldDimensions" (a tuple of booleans) into horizontal and vertical "field dimensions", similarly as we did it for extents and offsets.

### Resolves / Enhances

Fixes #415

### Dependencies

This PR is independent

### CODE REVIEW

**Comment by mroethlin** [comment]
launch jenkins

File: dawn/src/dawn/AST/Offsets.h [hunk 1]
```diff
@@ -24,6 +24,7 @@ namespace dawn::ast {
 
 class Offsets;
 
+// TODO: move those?
```
Comment (havogt) [inline dawn/src/dawn/AST/Offsets.h]: I'd just create a new header.

File: dawn/src/dawn/CodeGen/CodeGen.h [hunk 2]
```diff
@@ -74,7 +74,8 @@ class CodeGen {
 
   static std::string getStorageType(const sir::Field& field);
   static std::string getStorageType(const iir::Stencil::FieldInfo& field);
-  static std::string getStorageType(Array3i dimensions);
+  static std::string getStorageType(const sir::FieldDimension& dimensions);
+  static std::string getStorageType(const Array3i& dimensions);
```
Comment (havogt) [inline dawn/src/dawn/CodeGen/CodeGen.h]: Is this old one still used? Should be removed?

File: dawn/src/dawn/IIR/StencilMetaInformation.h [hunk 3]
```diff
@@ -154,7 +154,11 @@ class StencilMetaInformation : public NonCopyable {
 
   void insertAccessOfType(FieldAccessType type, int AccessID, const std::string& name);
 
-  int addField(FieldAccessType type, const std::string& name, const Array3i fieldDimensions,
+  int addField(FieldAccessType type, const std::string& name,
+               const sir::FieldDimension& fieldDimensions,
+               ast::Expr::LocationType = ast::Expr::LocationType::Cells);
+
+  int addField(FieldAccessType type, const std::string& name, const Array3i& fieldDimensions,
```
Comment (havogt) [inline dawn/src/dawn/IIR/StencilMetaInformation.h]: same question: is the old one still needed?

File: dawn/src/dawn/SIR/SIR.h [hunk 4]
```diff
@@ -15,6 +15,7 @@
 #ifndef DAWN_SIR_SIR_H
 #define DAWN_SIR_SIR_H
 
+#include "dawn/AST/Offsets.h"
```
Comment (havogt) [inline dawn/src/dawn/SIR/SIR.h]: remove once tags are moved

File: dawn/src/dawn/SIR/SIR.h [hunk 5]
```diff
@@ -149,15 +150,74 @@ struct StencilFunctionArg {
...
+      : maskI_(dimi), maskJ_(dimj), maskK_(dimk) {}
+};
+
+class FieldDimension {
+  std::unique_ptr<FieldDimensionImpl> impl_;
+
+public:
+  FieldDimension() = delete;
+  ~FieldDimension() = default;
+  FieldDimension(dawn::ast::cartesian_, std::array<bool, 3> mask)
```
Comment (havogt) [inline dawn/src/dawn/SIR/SIR.h]: Do we need both constructors?

Comment (havogt) [reply dawn/src/dawn/SIR/SIR.h]: Actually I think both constructors should be removed.
@lukasm91 why do we need this constructor with tag in Offset/Extent?

File: dawn/src/dawn/Serialization/SIRSerializer.cpp [hunk 6]
```diff
@@ -258,8 +258,10 @@ static std::shared_ptr<sir::Field> makeField(const dawn::proto::statements::Fiel
     if(fieldProto.field_dimensions().size() > 3)
       throwException("field_dimensions");
 
-    std::copy(fieldProto.field_dimensions().begin(), fieldProto.field_dimensions().end(),
-              field->fieldDimensions.begin());
+    field->fieldDimensions = sir::FieldDimension(
+        dawn::ast::cartesian, std::array<bool, 3>({(bool)fieldProto.field_dimensions().at(0),
+                                                   (bool)fieldProto.field_dimensions().at(1),
+                                                   (bool)fieldProto.field_dimensions().at(2)}));
```
Comment (havogt) [inline dawn/src/dawn/Serialization/SIRSerializer.cpp]: With the comment above, I think it should be
```
    field->fieldDimensions = sir::CartesianFieldDimension(
       (bool)fieldProto.field_dimensions().at(0),
                                                   (bool)fieldProto.field_dimensions().at(1),
                                                   (bool)fieldProto.field_dimensions().at(2));
```

File: dawn/src/dawn/SIR/SIR.h [hunk 7]
```diff
@@ -149,15 +150,74 @@ struct StencilFunctionArg {
...
+  bool maskI_, maskJ_, maskK_;
+  std::unique_ptr<FieldDimensionImpl> cloneImpl() const override {
+    return std::make_unique<CartesianFieldDimension>(maskI_, maskJ_, maskK_);
+  }
+
+public:
+  bool I() const { return maskI_; }
+  bool J() const { return maskJ_; }
+  bool K() const { return maskK_; }
+  ~CartesianFieldDimension() = default;
```
Comment (lukasm91) [inline dawn/src/dawn/SIR/SIR.h]: not needed

File: dawn/src/dawn/SIR/SIR.h [hunk 8]
```diff
@@ -149,15 +150,74 @@ struct StencilFunctionArg {
...
+  CartesianFieldDimension(std::array<bool, 3> mask)
```
Comment (lukasm91) [inline dawn/src/dawn/SIR/SIR.h]: explicit

File: dawn/src/dawn/SIR/SIR.h [hunk 9]
```diff
@@ -149,15 +150,74 @@ struct StencilFunctionArg {
...
+  CartesianFieldDimension(bool dimi, bool dimj, bool dimk)
+      : maskI_(dimi), maskJ_(dimj), maskK_(dimk) {}
+};
+
+class FieldDimension {
+  std::unique_ptr<FieldDimensionImpl> impl_;
+
+public:
+  FieldDimension() = delete;
+  ~FieldDimension() = default;
```
Comment (lukasm91) [inline dawn/src/dawn/SIR/SIR.h]: both lines not needed

**Comment by mroethlin** [comment]
launch jenkins

**Comment by mroethlin** [comment]
launch jenkins

**Comment by mroethlin** [comment]
launch jenkins

**Comment by mroethlin** [comment]
launch jenkins

**Comment by mroethlin** [comment]
launch jenkins

**Comment by mroethlin** [comment]
launch jenkins

**Comment by mroethlin** [comment]
launch jenkins

File: dawn/src/dawn/SIR/SIR.h (hunk 5 above)
Comment (havogt) [reply dawn/src/dawn/SIR/SIR.h]: We resolved the second question. The first still remains. Maybe we should remove the second ctor.

**Comment by havogt** [comment]
launch jenkins

**Comment by mroethlin** [comment]
launch jenkins

**Comment by mroethlin** [comment]
launch jenkins

**Comment by mroethlin** [comment]
launch jenkins

**Pull request merged**